File: src/Solutions/Euler217.hs
module Solutions.Euler217 where
import Solutions.Euler171(K, V, M, plusV, transform)
import Control.Monad
import Control.Monad.Memo
import Data.Array
import qualified Data.Map.Strict as Map

-- How long number has to be, what sum of digits must it have, can we have leading zero?
-- type K = (Integer, Integer, Bool)
-- type V = (Integer, Integer) -- How many numbers are there and what is their sum?
-- type M = Map.Map K V

initMap :: M
initMap = Map.fromList $ [((1, x, False), (if x > 0 then 1 else 0, x)) | x <- [0..9]] ++ [((1, x, True), (1, x)) | x <- [0..9]]

compute :: K -> Memo K V V
compute args@(numLength, digitsSum, zeroAllowed) =
    if digitsSum < 0 || numLength < 1
        then return (0, 0)
        else foldr plusV (0, 0) <$> forM [if zeroAllowed then 0 else 1..9]
                (\d -> transform d numLength <$> memo compute (numLength - 1, digitsSum - d, True))

doComputation :: Memo K V ()
doComputation = forM_ [2..23] $ \x -> forM_ [0..9 * x] $ \y -> memo compute (x, y, True) >> memo compute (x, y, False)

isValidI :: (K, V) -> Bool
isValidI ((a, b, _), _) = 1 <= a && a <= 23 && 0 <= b && b <= 207

arr :: Array K V
arr = array ((1, 0, False), (23, 207, True)) $ filter isValidI $ Map.assocs $ snd $ runMemo doComputation initMap

balanced' :: Integer -> Integer -> Integer
balanced' n digitsSum
    | even n = let (upperCount, upper) = arr ! (n `div` 2, digitsSum, False)
                   (lowerCount, lower) = arr ! (n `div` 2, digitsSum, True)
               in lowerCount * upper * 10^(n `div` 2) + upperCount * lower
    | otherwise = let (upperCount, upper) = arr ! (n `div` 2, digitsSum, False)
                      (lowerCount, lower) = arr ! (n `div` 2, digitsSum, True)
                   in 10 * (lowerCount * upper * 10^(n `div` 2 + 1) + upperCount * lower) + lowerCount * upperCount * 45 * 10^(n `div` 2)

balanced :: Integer -> Integer
balanced n = if n == 1 then 45 else sum $ map (balanced' n) [1..9 * (n `div` 2)]

euler217 :: IO String
euler217 = return $ show $ (`mod` (3^15)) $ sum $ map balanced [1..47]

File: rust/src/solutions/euler217.rs
use cached::proc_macro::cached;

// Input: How long number has to be,
//        what sum of digits must it have,
//        can we have leading zero?
// Output: How many numbers are there and what is their sum (mod)?
fn m(x: u128) -> u128 {
    x % 3u128.pow(15)
}

fn p(n: u32) -> u128 {
    m(10u128.pow(n))
}

#[cached]
fn count_nums(len: u32, sum_of_digits: u128, leading_zero_allowed: bool) -> (u128, u128) {
    if len == 1 {
        if sum_of_digits > 9 {
            (0, 0)
        } else if leading_zero_allowed {
            (1, sum_of_digits)
        } else {
            (if sum_of_digits > 0 { 1 } else { 0 }, sum_of_digits)
        }
    } else {
        (if leading_zero_allowed { 0 } else { 1 }..10)
            .filter_map(|d| {
                if sum_of_digits < d {
                    None
                } else {
                    let (how_many, total_sum) = count_nums(len - 1, sum_of_digits - d, true);
                    Some((how_many, total_sum + d * how_many * p(len - 1)))
                }
            })
            .fold((0, 0), |(a, b), (c, d)| (m(a + c), m(b + d)))
    }
}

fn balanced(len: u32) -> u128 {
    fn balanced_(len: u32, sum_of_digits: u128) -> u128 {
        let n = len / 2;
        let (upper_count, upper) = count_nums(n, sum_of_digits, false);
        let (lower_count, lower) = count_nums(n, sum_of_digits, true);
        if len % 2 == 0 {
            lower_count * upper * p(n) + upper_count * lower
        } else {
            10 * (lower_count * upper * p(n + 1) + upper_count * lower)
                + lower_count * upper_count * 45 * p(n)
        }
    }
    if len == 1 {
        45
    } else {
        (1..=9 * (len as u128 / 2))
            .map(|sum_of_digits| balanced_(len, sum_of_digits))
            .fold(0, |x, y| m(x + y))
    }
}

pub fn euler217() -> String {
    (1..=47).map(balanced).fold(0, |x, y| m(x + y)).to_string()
}
